src/common/ad_tx.c

```diff
@@ -835,40 +982,57 @@ umo_tx_add(struct umem_instance *umm, umem_off_t umoff, uint64_t offset, size_t
 	void			*ptr;
 
 	D_ASSERT(offset == 0);
-	D_ASSERTF(ad_tx_stage(tx) == UMEM_STAGE_WORK,
-		  "TX "DF_U64", bad stage %d\n", ad_tx_id(tx), ad_tx_stage(tx));
 	ptr = blob_addr2ptr(blob, umoff);
-	return ad_tx_snap(tx, ptr, size, AD_TX_UNDO);
+	return tx_add_internal(tx, ptr, size, AD_TX_UNDO | AD_TX_REDO);
 }
 
 static int
 umo_tx_xadd(struct umem_instance *umm, umem_off_t umoff, uint64_t offset, size_t size,
 	    uint64_t flags)
 {
-	/* NOOP for UMEM_XADD_NO_SNAPSHOT */
-	if (flags & UMEM_XADD_NO_SNAPSHOT)
-		return 0;
+	struct ad_tx		*tx = tx_get();
+	struct ad_blob_handle	 bh = umm2ad_blob_hdl(umm);
+	struct ad_blob		*blob = bh.bh_blob;
+	void			*ptr;
+	uint32_t		 ad_flags = 0;
+
+	if (!(flags & UMEM_XADD_NO_SNAPSHOT))
+		ad_flags |= AD_TX_UNDO;
+	if (!(flags & UMEM_FLAG_NO_FLUSH))
+		ad_flags |= AD_TX_REDO;
```

Comment:
if using NO_FLUSH does not make sense, we probably can assert it. let's see @jolivier23 's comment, thx

Reply:
As far as I can tell, this flag is only used in the PMDK based allocator code so probably a better question for @sherintg .

Valgrind only uses NO_SNAPSHOT though in our case, I think we will want to mark such pages as dirty even if we don't support rollback.  I'm thinking of cases like the latest write on objects where we don't care if it doesn't get rolled back to an older value but where we definitely want it flushed on commit.

Reply:
I agree there is no valid use case for this flag with the blob backed memory implementation. For DAV, it implies that this address range will **not** go into the wal redo log unless some other operation like the atomic copy persists it. Otherwise, this address range will be persisted during the next checkpoint operation. I see the same behavior defined for ad allocator aswell.

Reply:
If it's not a valid use case, I think we'd return error (or assert), that could help us to catch potential bugs.

Reply:
ok, I'll change to assert if caller pass in NO_FLUSH flag, in next patch or in this patch if need to refresh later. Thanks you all for the info.